pkg/storage/azure/azure.go

```diff
@@ -0,0 +1,506 @@
+package azure
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"reflect"
+	"regexp"
+
+	"github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2019-04-01/storage"
+	"github.com/Azure/azure-storage-blob-go/azblob"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/golang/glog"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/client-go/rest"
+
+	operatorapiv1 "github.com/openshift/api/operator/v1"
+
+	imageregistryv1 "github.com/openshift/cluster-image-registry-operator/pkg/apis/imageregistry/v1"
+	regopclient "github.com/openshift/cluster-image-registry-operator/pkg/client"
+	"github.com/openshift/cluster-image-registry-operator/pkg/clusterconfig"
+	"github.com/openshift/cluster-image-registry-operator/pkg/parameters"
+	"github.com/openshift/cluster-image-registry-operator/pkg/storage/util"
+)
+
+const (
+	blobFormatString = `https://%s.blob.core.windows.net`
+
+	storageExistsReasonNotConfigured     = "StorageNotConfigured"
+	storageExistsReasonConfigError       = "ConfigError"
+	storageExistsReasonUserManaged       = "UserManaged"
+	storageExistsReasonAzureError        = "AzureError"
+	storageExistsReasonContainerNotFound = "ContainerNotFound"
+	storageExistsReasonContainerExists   = "ContainerExists"
+	storageExistsReasonContainerDeleted  = "ContainerDeleted"
+	storageExistsReasonAccountDeleted    = "AccountDeleted"
+)
+
+var (
+	nonAlnumRe = regexp.MustCompile(`[^0-9a-z]`)
+)
+
+type errNameNotAvailable struct {
+	AccountName string
+	Message     string
+}
+
+func (e *errNameNotAvailable) Error() string {
+	return fmt.Sprintf("storage account name %s is not available: %s", e.AccountName, e.Message)
+}
+
+type errDoesNotExist struct {
+	Err error
+}
+
+func (e *errDoesNotExist) Error() string {
+	return e.Err.Error()
+}
+
+// generateAccountName returns a name that can be used for an Azure Storage
+// Account. Storage account names must be between 3 and 24 characters in
+// length and use numbers and lower-case letters only.
+func generateAccountName(infrastructureName string) string {
+	prefix := nonAlnumRe.ReplaceAllString(infrastructureName, "")
+	if prefix == "" {
+		prefix = "imageregistry"
+	}
+	if len(prefix) > 24-5 {
+		prefix = prefix[:24-5]
+	}
+	return prefix + rand.String(5)
+}
+
+func createStorageAccount(storageAccountsClient storage.AccountsClient, resourceGroupName, accountName, location string) error {
+	glog.Infof("attempt to create azure storage account %s (resourceGroup=%q, location=%q)...", accountName, resourceGroupName, location)
+
+	ctx := context.TODO()
```

Comment:
We may want to have a single Context for the entire reconciliation cycle. In this case we should get context from the caller. And `context.TODO()` means that this context eventually should be replaced by the context from the caller.

Otherwise TODO and Background have the same meaning. So if you want, I can use Background here.

Reply:
keep TODO for now. We can come back later on this if needed.